### PR TITLE
Adding documentation builds

### PR DESCRIPTION
**Pull request by greglucas**
# Change Summary

## Overview

This adds sphinx documentation that can be built for easier exploration of API functionality. Based off of the workflow for this project, with an example of what this will look like when deployed: https://swxtrec.github.io/pymsis/

Future work:
* Add a GitHub deploy token so our release mechanism via GitHub Pages can work
* Set up GitHub pages branch as an orphan that we can deploy this into

## New Dependencies
sphinx, pydata-sphinx-theme

## New Files

Many under the `docs/` folder to start the basic structure

## Updated Files

- __init__.py
   - Added a version number, we should edit this later to be automatically done, but at least put something in for now

- pyproject.toml
   - Added a few new dependencies and a `[doc]` group.

## Testing

Current testing is done by just building the docs on PR and uploading the artifact that can be downloaded and inspected. It would be good to post the deployed artifact somewhere for easier viewing, which we can automate via CircleCI. But that can come in follow-up PRs if people want.

### CODE REVIEW

**Comment by laspsandoval** [comment]
Nice! I really like the automated documentation approach. A few things:

I think we agreed PRs should be created to merge to dev branch

We might start making .md files in the docs directory to provide folks info on how to use the repo. For example add this to describe the build process for sphinx:
1. poetry install --extras "doc"
2. cd docs/source
3. poetry run sphinx-build . _build
4. open _build/index.html

Do we want to create documentation on Github Pages and/or Confluence? Also, Matt Watwood has done this for Libera in Confluence so before we reinvent the wheel we might take a look at what was done.

**Comment by laspsandoval** [comment]
Here's an example of the markdown file for Libera, which is probably far more than we need, but may be a helpful template. 

[documentation-sphinx.md](https://github.com/IMAP-Science-Operations-Center/imap_processing/files/11958436/documentation-sphinx.md)

**Comment by maxinelasp** [comment]
Is this for external documentation? or internal? So, is this intended to replace confluence or the github wiki, or does it operate as more of a readthedocs type thing?

**Comment by greglucas** [comment]
> Is this for external documentation? or internal? So, is this intended to replace confluence or the github wiki, or does it operate as more of a readthedocs type thing?

I view this as documenting the **code** and a little less focused on all of the confluence stuff. I think since this is a public repo, the docs should be public as well. We can certainly push to readthedocs as well if we'd prefer that route instead of GitHub Pages, I'm indifferent.

> I think we agreed PRs should be created to merge to dev branch

Good catch! dev is unfortunately behind right now, but I can rebase and point to that once the merge-back PR is in.

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+2. Build the html documents under the `docs` directory
+
+Using poetry:
+
+```bash
+# Install the optional documentation dependencies
+poetry install --extras "doc"
+# Make the html documentation
+cd docs/source
+poetry run sphinx-build . build
```
Comment (tech3371) [inline docs/README.md:15]: What is sphinx-build? What does that argument mean?

File: docs/make.bat [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+)
+set SOURCEDIR=source
+set BUILDDIR=build
+
+if "%1" == "" goto help
+
+%SPHINXBUILD% >NUL 2>NUL
+if errorlevel 9009 (
+	echo.
+	echo.The 'sphinx-build' command was not found. Make sure you have Sphinx
```
Comment (tech3371) [inline docs/make.bat:18]: Ah I see. This answered my previous question. Is there way to add this to pyproject.toml? There may be way to install it using PyPI: https://www.sphinx-doc.org/en/master/usage/installation.html#installation-from-pypi

File: docs/source/conf.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+This file only contains a selection of the most common options. For a full
+list see the documentation:
+https://www.sphinx-doc.org/en/master/usage/configuration.html
+"""
+
+# -- Path setup --------------------------------------------------------------
+
+import sys
+
+# If extensions (or modules to document with autodoc) are in another directory,
```
Comment (tech3371) [inline docs/source/conf.py:13]: These documentation are really helpful. Thank you!

File: docs/source/development/index.rst [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    # Update the code on a feature branch
+    git checkout -b my-cool-feature
+
+    # Run the tests
+    pytest
+
+    # Commit the changes and push to your remote repository
+    git add my-file
+    git commit
+    git push -u origin my-cool-feature
```
Comment (tech3371) [inline docs/source/development/index.rst:33]: This looks like a great content to add to style guide. Or style guild could be included in this automated doc generator.

File: docs/source/index.rst [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+contribute your expertise, and help unlock the mysteries of the interstellar
+medium and particle acceleration in our galaxy. Together, we can advance our
+knowledge of the cosmos and pave the way for future discoveries in space
+exploration.
+
+The [IMAP Website](https://imap.princeton.edu/) contains information about the
+entire mission, including the science goals, the spacecraft, and the instruments.
+The Science Operations Center (SOC) is responsible for the data processing on the
+IMAP mission and being developed at
+[LASP (Laboratory of Atmospheric and Space Physics)](https://lasp.colorado.edu/).
```
Comment (tech3371) [inline docs/source/index.rst:33]: nice!

File: pyproject.toml [hunk 6]
```diff
@@ -31,15 +31,17 @@ classifiers = [
...
-
-
+# Optional dependencies
 black = {version="^23.1.0", optional=true}
 pre-commit = {version="^3.3.3,<4", optional=true}
-ruff = {version="^0.0.253", optional=true}
+pydata-sphinx-theme = {version="*", optional=true}
 pytest = {version=">=6.2.5,<7", optional=true}
+ruff = {version="^0.0.253", optional=true}
+sphinx = {version="*", optional=true}
```
Comment (tech3371) [inline pyproject.toml:40]: I see. You did add it to the toml file. Thank you!

**Review by tech3371 — APPROVED** [review]

**Comment by maxinelasp** [comment]
> I view this as documenting the code and a little less focused on all of the confluence stuff. I think since this is a public repo, the docs should be public as well. 

Sounds good to me! I'd like if we could define what kind of documentation we'd like to see (more templates, anyone?) but I think this looks good! Maybe we can require documentation as part of a PR, same as tests?

**Review by maxinelasp — APPROVED** [review]

File: docs/source/development/index.rst (hunk 4 above)
Comment (greglucas) [reply docs/source/development/index.rst:33]: Yep, I agree, I think the style guide should either go into the docs, or vice-versa the style guide point to the docs.

**Pull request merged**